src/rp2_common/hardware_gpio/include/hardware/gpio.h

```diff
@@ -365,46 +375,243 @@ enum gpio_drive_strength gpio_get_drive_strength(uint gpio);
+ * \note The IO IRQs are independent per-processor. This configures affects the processor that calls the function.
  *
+ * \param gpio GPIO number
+ * \param events Which events will cause an interrupt. See \ref gpio_irq_level for details.
+ * \param enabled Enable or disable flag
+ * \param callback user function to call on GPIO irq. if NULL, the callback is removed
  */
 void gpio_set_irq_enabled_with_callback(uint gpio, uint32_t events, bool enabled, gpio_irq_callback_t callback);
 
-/*! \brief Enable dormant wake up interrupt for specified GPIO
+/*! \brief Enable dormant wake up interrupt for specified GPIO and events
  *  \ingroup hardware_gpio
  *
  * This configures IRQs to restart the XOSC or ROSC when they are
  * disabled in dormant mode
  *
  * \param gpio GPIO number
- * \param events Which events will cause an interrupt. See \ref gpio_set_irq_enabled for details.
+ * \param events Which events will cause an interrupt. See \ref gpio_irq_level for details.
  * \param enabled Enable/disable flag
  */
 void gpio_set_dormant_irq_enabled(uint gpio, uint32_t events, bool enabled);
 
-/*! \brief Acknowledge a GPIO interrupt
+/*! \brief Return the current interrupt status (pending events) for the given GPIO
  *  \ingroup hardware_gpio
  *
  * \param gpio GPIO number
- * \param events Bitmask of events to clear. See \ref gpio_set_irq_enabled for details.
-  *
+ * \return events Bitmask of events that are currently pending for the GPIO. See \ref gpio_irq_level for details.
+ * \sa gpio_acknowledge_irq
+ */
+static inline uint32_t gpio_get_irq_events(uint gpio) {
+    io_irq_ctrl_hw_t *irq_ctrl_base = get_core_num() ?
+                                      &iobank0_hw->proc1_irq_ctrl : &iobank0_hw->proc0_irq_ctrl;
+    io_ro_32 *status_reg = &irq_ctrl_base->ints[gpio >> 3u];
+    return (*status_reg >> (4 * (gpio & 7u))) & 0xfu;
+}
+
+/*! \brief Acknowledge a GPIO interrupt for the specified events
+ *  \ingroup hardware_gpio
+ *
+ * \param gpio GPIO number
+ * \param events Bitmask of events to clear. See \ref gpio_irq_level for details.
  */
 void gpio_acknowledge_irq(uint gpio, uint32_t events);
 
+/*! \brief Adds a raw GPIO IRQ handler for the specified GPIOs on the current core
+ *  \ingroup hardware_gpio
+ *
+ * In addition to the default mechanism of a single GPIO IRQ event callback per core (see \ref gpio_set_irq_callback),
+ * it is possible to add explicit GPIO IRQ handlers which are called independent of the default callback. The order
+ * relative to the default callback can be controller via the order_priority parameter. (The default callback has the priority
+ * \ref GPIO_CALLBACK_IRQ_ORDER_PRIORITY)
+ *
+ * This method adds such a explicit GPIO IRQ handler, and disables the "default" callback for the specified GPIOs.
+ *
+ * A raw handler should check for whichever GPIOs and events it handles, and acknowledge them itself; it might look something like:
```

Comment:
This is probably a dumb question, but if the user has to manually check for GPIO numbers inside their callback-function themselves, what's the advantage of using these "raw handlers" over the "shared handler"? From looking at the implementation of `gpio_add_raw_irq_handler_with_order_priority_masked` it seems like if you added multiple raw-handlers (each with a separate callback), then _each_ of the callbacks would get called for *every* GPIO IRQ that's registered to be handled by a raw-handler?
So you could just do the same thing with the shared-handler, by calling a different function depending on the GPIO number that it was passed?
No doubt I'm missing something obvious / fundamental :confused:

Reply:
Ahhh, "libraries that need to hook a GPIO should add their own handler" was the bit I was missing, thanks :smiley: 
(is that worth clarifying in the doxygen?)

Reply:
i think these comments are outdated as I don't see the text "libraries that need to"